lib/pos_blockchain.go

```diff
@@ -57,66 +57,145 @@ func (bc *Blockchain) processBlockPoS(block *MsgDeSoBlock, currentView uint64, v
 	}
 	if !parentBlockNode.IsValidated() {
 		blk, err := GetBlock(parentBlockNode.Hash, bc.db, bc.snapshot)
 		if err != nil {
-			return false, false, nil, errors.Wrapf(err, "processBlockPoS: Problem getting parent block %v", parentBlockNode.Hash)
-		}
-		_, _, _, err = bc.processBlockPoS(blk, currentView, verifySignatures)
-		if err != nil {
-			return false, false, nil, errors.Wrapf(err, "processBlockPoS: Problem processing parent block %v", parentBlockNode.Hash)
+			// If we don't have the parent, we need to fetch it from the network.
+			// We won't handle that in here, so we should just store the block and return.
+			if storeInBlockIndexErr := bc.storeBlockInBlockIndex(block); storeInBlockIndexErr != nil {
+				return errors.Wrapf(err, "validateBlockPoS: Problem adding block to block index: %v", storeInBlockIndexErr)
+			}
+			return errors.Wrapf(err, "validateBlockPoS: Problem getting parent block %v", parentBlockNode.Hash)
+		}
+		// TODO: I hate how deeply nested this is right now, but we'll refactor to make it nicer.
+		if err = bc.validateBlockPoS(blk); err != nil {
+			// Okay here's where it gets a little tricky. We pull the parent BlockNode from the BlockIndex. If the
+			// parent has a status of ValidateFailed, then we know we store this block as ValidateFailed. If the parent
+			// is not ValidateFailed, we ONLY store the block and move on. We don't want to store it as ValidateFailed
+			// because we don't know if it's actually invalid. It could be that we hit some intermittent error when
+			// trying to validate its parent, such as an intermittent badger error.
+			parentBlockNode = bc.blockIndex[*block.Header.PrevBlockHash]
+			if parentBlockNode.IsValidateFailed() {
+				if storeInBlockIndexErr := bc.storeValidateFailedBlockInBlockIndex(block); storeInBlockIndexErr != nil {
+					return errors.Wrapf(err, "validateBlockPoS: Problem adding validate failed block to block index: %v", storeInBlockIndexErr)
+				}
+			} else {
+				if storeInBlockIndexErr := bc.storeBlockInBlockIndex(block); storeInBlockIndexErr != nil {
+					return errors.Wrapf(err, "validateBlockPoS: Problem adding block to block index: %v", storeInBlockIndexErr)
+				}
+			}
+			return errors.Wrapf(err, "validateBlockPoS: Problem processing parent block %v", parentBlockNode.Hash)
 		}
 	}
 
-	// At this point, we know that all the ancestors of this block have been successfully
-	// validated. That means we can now proceed with attempting to validate THIS block.
-
-	// Do all the sanity checks on the block.
-	// TODO: Check if err is for view > latest committed block view and <= latest uncommitted block.
-	// If so, we need to perform the rest of the validations and then add to our block index.
-	if err = bc.validateDeSoBlockPoS(block); err != nil {
-		// TODO: If validations fail, do we want to even bother storing the block? Are there any validations that
+	if err := bc.validateDeSoBlockPoS(block); err != nil {
```

Comment:
I realize you're still brainstorming names from the block validations. Consider calling this something like `isValidPoSBlock()` or `isValidBlockPoS()` so that it's clear the function just checks if the block is correct without any other side effects
```suggestion
	if err := bc.isValidPoSBlock(block); err != nil {
```